examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/user_data.c

```diff
@@ -0,0 +1,361 @@
+ *       datasctructure when the data in flash is valid. A simple
+ *       (user_data_t*)FLASH_USER_DATA_BASE will give you a pointer to the data in flash.
+ *       It cannot be written to directly (flash erase and program are required for
+ *       that), but it can be CRC-validated and accessed from anywhere in the codebase
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <stdint.h>
+#include "stm32f4xx.h"
+#include "stm32f4xx_hal_conf.h"
+#include "user_data.h"
+
+#define CRC_REGISTER_SIZE sizeof( int32_t )
+static int8_t calculate_checksum( user_data_t* user_data, int32_t* out_checksum );
+
+#define FLASH_USER_DATA_SECTOR FLASH_SECTOR_TOTAL - 1
+#define FLASH_VOLTAGE_RANGE FLASH_VOLTAGE_RANGE_1
+#define FLASH_USER_DATA_ALLOCATED_SIZE 0xfff
+#define FLASH_USER_DATA_BASE ( FLASH_END - FLASH_USER_DATA_ALLOCATED_SIZE )
+static int8_t erase_flash_sector( void );
+static int8_t program_flash( user_data_t* user_data );
+
+/******************************************************************************
+*                           Workflow Implementation                           *
+******************************************************************************/
+
+/**
+ * @brief  Verify this device is compatible with the hardcoded flash
+ *         sector/address specifications
+ * @param   None
+ * @retval 0 Success
+ * @retval <0 Error: Something is probably misconfigured in the source code
+ */
+int8_t user_data_flash_init( void )
+{
+    assert( 0 == sizeof( user_data_t ) % sizeof( int32_t ) ); /* For program_flash() */
+    assert( 0 == USER_DATA_STR_SIZE % CRC_REGISTER_SIZE );
+    if ( !IS_FLASH_SECTOR( FLASH_USER_DATA_SECTOR ) )
+    {
+        printf( "\r\n>> Flash init [ERROR] Sector [%d] is incompatible with this device",
+                FLASH_USER_DATA_SECTOR );
+        return -1;
+    }
+    if ( sizeof( user_data_t ) > FLASH_USER_DATA_ALLOCATED_SIZE )
+    {
+        printf( "\r\n>> Flash init [ERROR] Allocated flash meomry not enough" );
+        return -1;
+    }
+
+    return 0;
+}
+
+/**
+ * @brief  Copy the Flash area reserved for user data into *dst so it can be
+ *         manipulated
+ * @param  *dst is the pre-allocated pointer where we want to copy the user data
+ * @retval <0: Error, >=0: OK
+ */
+int8_t user_data_copy_from_flash( user_data_t* dst )
+{
+    if ( NULL == dst )
+    {
+        return -1;
+    }
+
+    memset( dst, 0x00, sizeof( user_data_t ) );
+    memcpy( dst, ( user_data_t* )FLASH_USER_DATA_BASE, sizeof( user_data_t ) );
+    return 0;
+}
+
+/**
+ * @brief  This function calculates and sets the CRC checksum of the user data
+ *         we'd like to save, erases the relevant Fash sector and stores the new
+ *         data
+ * @param  *src is the datastructure we'd like to save to flash. src->crc_checksum
+ *         will be overwritten
+ * @retval <0: Error, >=0: OK
+ */
+/* TODO: This execution branch should run entirely from volatile RAM if possible */
```

Comment:
Well, I'd love to revisit this if we have the time. I think it's an implementation with good business value (in a production environment it would be used to store multiple datastructures in a memory area reserved at the linker script, not to separate code from user data, but the code is the same).

That being said, I can certainly change the comments to `WARNING`s if we're not gonna have the time to implement this.